### ISSUE TITLE
Flavor-conditional asset bundling does not work when declared within a direct main path dependencies

### ISSUE THREAD

**Issue by alestiago** [flutter/flutter#155755]
### Description

Flutter 3.22 introduced “Flavor-conditional asset bundling”, read more about them in the [“Conditionally bundling assets based on flavor” documentation](https://docs.flutter.dev/deployment/flavors#conditionally-bundling-assets-based-on-flavor).

When declaring assets outside of the main application package and defining flavors for them the main application bundle does not account for these assets as expected.

### Steps to reproduce

1. Create a new flutter application for Android:
```sh
flutter create my_app --platforms android
```
2. Create a new flutter package:
```sh
flutter create my_app/packages/my_assets --template package
```

3. Add “package:my_assets” as a direct main dependency on “package:my_app”:
```sh
flutter pub add 'my_assets:{"path":"my_app/packages/my_assets"}' --directory my_app
```

4. Create new assets directories  in “package:my_app” and “package:my_assets”:
```sh
mkdir my_app/assets
mkdir my_app/packages/my_assets/assets
```

5. Save new assets on each assets directory, such that:
```txt
my_app/
├─ assets/
│  ├─ my_app_logo.png
│  ├─ my_app_logo-dev.png
│  ├─ my_app_logo-prod.png
├─ packages/
│  ├─ my_assets/
│  │  ├─ assets/
│  │  │  ├─ my_assets_logo-dev.png
│  │  │  ├─ my_assets_logo-prod.png
│  │  │  ├─ my_assets_logo.png
```

6. Declare the assets as flavored assets in “package:my_app”:

```yaml
# file: my_app/pubspec.yaml

# ...

flutter:
 assets:
   - assets/my_app_logo.png
   - path: assets/my_app_logo-dev.png
     flavors:
       - dev
   - path: assets/my_app_logo-prod.png
     flavors:
       - prod
```

7. Declare the assets as flavored assets in “package:my_assets”:

```yaml
# file: my_app/packages/my_assets/pubspec.yaml

# ...

flutter:
 assets:
   - assets/my_assets_logo.png
   - path: assets/my_assets_logo-dev.png
     flavors:
       - dev
   - path: assets/my_assets_logo-prod.png
     flavors:
       - prod
```

8. Define “dev” and “prod” [Android flavors](https://docs.flutter.dev/deployment/flavors#using-flavors-in-android) in “my_app/android/app/build.gradle”:

```gradle
android {
   // ...

   flavorDimensions "default"

   productFlavors {
       dev {
           dimension "default"
           resValue "string", "app_name", "dev flavor example"
           applicationIdSuffix ".dev"
       }
       prod {
           dimension "default"
           resValue "string", "app_name", "prod flavor example"
           applicationIdSuffix ".prod"
       }
   }
}
```

9. Bundle the main application as an APK:

```sh
flutter build apk --flavor dev
```

10. Analyze the generated APK using [Android Studio’s APK Analyzer](https://developer.android.com/studio/debug/apk-analyzer).

The built APK is expected to be at: my_app/build/app/outputs/flutter-apk/app-dev-release.apk

### Expected results

**_my_assets_logo-dev.png_** should have also been bundled in the APK but it was not.

### Actual results

The following assets were included in the APK:
- my_app_logo.png
- my_app_logo-dev.png
- my_assets_logo.png

### Code sample

Not applicable, see reproductive steps instead.

### Screenshots or Video

<details open>
<summary> Actual APK Analyzer results </summary>

<img width="895" alt="apk-analyzer" src="https://github.com/user-attachments/assets/5f587308-4f7d-4558-bced-e4377ffcf2e8">

</details>

### Logs

Not applicable.

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.24.3, on macOS 14.7 23H124 darwin-arm64, locale en-US)
    • Flutter version 3.24.3 on channel stable at /Users/alestiago-nbcu/Developer/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (2 weeks ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/alestiago-nbcu/Library/Android/sdk
    • Platform android-35, build-tools 35.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)

[✓] VS Code (version 1.93.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.96.0

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.7 23H124 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.7 23H124 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 129.0.6668.60

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by huycozy** [comment]
Hi @alestiago 
I think the problem comes from assets path. Per the example on the docs, you should create separate directories for each flavors, for eg:

```yaml
flutter:
 assets:
   - assets/common/
   - path: assets/dev/
     flavors:
       - dev
   - path: assets/prod/
     flavors:
       - prod
```

```
├── assets
│   ├── common
│   │   └── flutter.jpeg
│   ├── dev
│   │   └── flutter.jpeg
│   └── prod
│       └── flutter.jpeg
```

And similar assets declaration in the package. Can you try to see if it works?

**Comment by alestiago** [comment]
Hi @huycozy, thanks for chiming in. 

Your suggestion is invalid, it does not bundle the flavored assets when defined in a dependency package, see below details for additional information.

<details>

<summary> Additional information </summary>

For clarity, this was the new directory structure:

```txt
my_app/
├─ assets/
│  ├─ my_app_logo.png
│  ├─ my_app_logo-dev.png
│  ├─ my_app_logo-prod.png
├─ packages/
│  ├─ my_assets/
│  │  ├─ assets/
│  │  │  ├─ my_assets_logo.png
│  │  │  ├─ dev/
│  │  │  │  ├─ my_assets_logo-dev.png
│  │  │  ├─ prod/
│  │  │  │  ├─ my_assets_logo-prod.png
```

This was the new pubspec within the `package:my_assets`:

```yaml
# file: my_app/packages/my_assets/pubspec.yaml

flutter:
  assets:
    - assets/my_assets_logo.png
    - path: assets/dev/
      flavors:
        - dev
    - path: assets/prod/
      flavors:
        - prod
```

And this was the examined APK after being built with:

```sh
flutter build apk --flavor dev
```

<img width="895" alt="Screenshot 2024-09-27 at 09 15 36" src="https://github.com/user-attachments/assets/7860e253-ed09-4b7e-84e3-d60ba2def7fc">


</details>

Please do **NOTE**, that regardless if it can only accept directories in dependencies or not (which also failed) what has been described in the issue is clearly showing **inconsistent behavior** in pubspecs when using flavor-conditional bundling. The same syntax was used for the main application package and the main dependency package, and the conditional bundling only worked as expected for the main application package.

**Comment by huycozy** [comment]
Thanks for your response. I can also see the issue with my assets structure. When building apk with flavor, conditionally bundling does not work on package.

<details>
<summary>My assets structure</summary>

```console
.
├── assets
│   ├── common
│   │   └── flutter.jpeg
│   ├── dev
│   │   └── flutter-dev.jpeg
│   └── prod
│       └── flutter-prod.jpeg
├── packages
│   └── my_assets
│       ├── assets
│       │   ├── common
│       │   │   └── flutter.jpeg
│       │   ├── dev
│       │   │   └── flutter-dev.jpeg
│       │   └── prod
│       │       └── flutter-prod.jpeg
```
</details>

### Output APK analysis

![Screenshot 2024-09-30 at 12 41 00](https://github.com/user-attachments/assets/d0974dac-d73d-4be6-ab79-d1f390d98d48)

The complete sample project: [flutter-flavor-conditionally-bundling-assets](https://github.com/huycozy/flutter-flavor-conditionally-bundling-assets). Reproduced on Flutter master `3.26.0-1.0.pre.263`.

**Comment by bkonyi** [comment]
FYI @andrewkolos

**Comment by feduke-nukem** [comment]
I would like to point out that this issue is not specific only to Android, as it also arises on iOS, according to https://github.com/flutter/flutter/issues/155404.

**Comment by andrewkolos** [comment]
This fix should be available in the next Flutter release (3.27 as far as I'm aware), starting with the beta branch.
